Dev fancontrold

nunojsa opened the pull request:
Daemon to control the fan speed on adrv9009-zu11eg based on the temperature reported by the monitored devices. The temperature table used to control the pwm can be seen at [fan-control](https://wiki.analog.com/resources/fpga/docs/axi_fan_control.).

commodo commented on fancontrold.c:
```diff
@@ -0,0 +1,489 @@
+	}
+
+	/* lookup for the hwmon device */
+	if (hwmon_find(dev_name) < 0) {
```
why not just do `ret = hwmon_find()` and return ret if negative ?

nunojsa replied:
yes, i should have done that. Also in the `iio_devices_scan()` call.

commodo commented on fancontrold.c:
```diff
@@ -0,0 +1,489 @@
+
+static char *sysfs_read_attr(const char *path, const char *attr)
+{
+	char buf[NAME_MAX];
```
why not just pass `buf` as a param to `sysfs_read_attr()` ?
it removes some alloc + free code paths

commodo commented on fancontrold.c:
```diff
@@ -0,0 +1,489 @@
+	errno = 0;
+	temp = strtol(buf, &endptr, 10);
+	/* check possible failures */
+	if ((errno == ERANGE && (temp == LONG_MAX || temp == LONG_MIN))
```
checking errno seems enough, especially since it's cleared before strtol()

nunojsa replied:
To be honest, i just followed the man page of `strtol`.

commodo commented on fancontrold.c:
```diff
@@ -0,0 +1,489 @@
+	const char *name = (dev_name == NULL) ? DEFAULT_DEVNAME : dev_name;
+	DIR *hwmon;
+	struct dirent *dir;
+	(void)dev_name;
```
this looks redundant ?

nunojsa replied:
ups. `(void)dev_name;` is something that should not be there. It was just to shut up warnings during development (when i was actually not using this variable)

commodo commented on fancontrold.c:
```diff
@@ -0,0 +1,489 @@
+		if (!saved_iio) {
+			error("Failed to allocate memory %s\n",
+						strerror(errno));
+			ret = -1;
```
you need to `free(an_monitor.iio_devs)` in this err path

nunojsa replied:
Hmm, not really. We are fine because it will be freed in the `clean_up()` call afterwards. I kinda of preferred to have all the clean up in one place. But i don't feel to strong about this. If you prefer me o change and to do all the cleanings "in place" (and ditch the `clean_up()` function), i can change it...

commodo replied:
hmm, ok; 
works to leave it like this and go in `clean_up()`

commodo commented on fancontrold.c:
```diff
@@ -0,0 +1,489 @@
+	} else if (optind + 1 == argc)
+		dev_name = argv[optind];
+
+	openlog(NULL, 0, LOG_DAEMON);
```
maybe give the log a name;
`openlog("adrv9009-fan-control", 0, LOG_DAEMON);`

nunojsa replied:
Fair enough :smile:

commodo commented:
rest looks good

nunojsa committed:
fan-control: Added daemon to control fan speed

This is a system daemon which will control de fan speed depending on the
reported temperature. At this moment the monitored devices are under the
iio bus and are:
- adrv9009-phy;
- adrv9009-phy-b.
This daemon is intended to be used on the adrv9009-zu11eg SOM.

Signed-off-by: Nuno Sá <nuno.sa@analog.com>

nunojsa committed:
build: Add cmake based build system

Signed-off-by: Nuno Sá <nuno.sa@analog.com>

nunojsa committed:
build: Add support for init scripts

As a daemon, we want it to start at system boot. Hence, added support
for systemd, sysV and upstart init systems.

Signed-off-by: Nuno Sá <nuno.sa@analog.com>

nunojsa commented:
Changes in V2:

- Simplify error handling in `strtol` call;
- Remove irrelevant `(void)dev_name` line;
- Add a name to the log, in `openlog()`;
- re-work some error handling paths.
